PTI-272: Update epic import to handle more types better.

NickPhura opened the pull request:
## Changes
- Added some new constants files to hold some of the epic record type IDs (docType, milestone, etc), so that information is all in one place should it need updating.
- Added a new constants file for the supported schema names and their corresponding display name.  If we use that everywhere that a schemaName or schema display name is used, then updating them will be super easy (if needed).
- Updated the existing import code so that 1 instance of the import task kicks off a datasource (epic), which in turn runs all of its supported record types.
  - The majority of this logic is the same as before, the only difference is that epic-datasource.js now runs the import logic for each of its supported types itself, whereas before it only handled one type at a time, and so multiple calls to import-task were needed.

## Other changes
- Updated the 3-4 instances of "recordSubType" to match the ~30 instances of "recordSubtype" (lowercase t).  Why? Because I got stuck for 15 minutes wondering why that field wasn't being filled out in one of my importers, turns out the code I initially copied form used a capitol 'T', but the model I copied from had a lowercase 't'.

## _Update (Jan 31)_

I rebased on the latest master (first commit).
I addressed code review changes (second commit).

NickPhura commented on api/src/integrations/epic/epic-orders.js:
```diff
@@ -87,10 +87,10 @@ class EpicOrders {
 
     try {
-      const Order = mongoose.model('Order');
+      const Order = mongoose.model(RECORD_TYPE.Order._schemaName);
```
I know its not likely to change, but its generally good practice to avoid magic strings.  

It also means that when you reference that model name in multiple places, there is no possible way for them to be inconsistent, because they all reference the single instance of the string from the enum, so can save headaches around misspelling things, etc.

That enum also has other uses, currently it relates the schema name to the UI display name (since they aren't the same).

NickPhura commented on api/src/models/master/order.js:
```diff
@@ -14,7 +14,7 @@ module.exports = require('../../utils/model-schema-generator')(
     recordType: { type: String, default: '' },
     recordSubtype: { type: String, default: '' },
-    dateIssued: { type: Date, default: null },
+    dateIssued: { type: Date, default: new Date() },
```
Gotcha, will fix.

danieltruong commented on angular/projects/admin-nrpti/src/app/import/import.component.ts:
```diff
@@ -155,9 +155,6 @@ export class ImportComponent implements OnInit {
-      this.factoryService.startTask({ dataSourceType: 'epic', recordType: 'order' })
-    );
+    return this.factoryService.startTask({ dataSourceType: 'epic' });
   }
```
I think we should still have these split up and just combine the results in the front end. I know they say they wont want just order or inspections to run by themselves now, but let's say they changed their mind on that. Does this change stop us from having that ability?

NickPhura replied:
In its current form, they would not be able to update just one record type.

But, it would be easy enough to make a small change to accept an optional parameter to specify a type.  And then it would just update that 1 type.  Then it would be able to handle both cases.

NickPhura commented on api/src/integrations/integration-datasource-enum.js:
```diff
@@ -0,0 +1,13 @@
+  }
+});
+
+module.exports = INTEGRATION_DATASOURCE;
```
The idea is that when we get more datasources, they all live here.  Currently we only have 1.

Also similar to the schema enum, this relates the swagger datasource type string to the display string, since they might not match (or might contain spaces, etc).

And also, its an alternative to big switch statements, because now you can just do:

```
const dataSourceType = INTEGRATION_DATASOURCE[swagger.stuff.datasourceType]
```

danieltruong replied:
If this just replaces a switch statement, isn't this more complicated and harder to maintain then a single switch statement kept within the import-task file?

NickPhura replied:
You could technically do everything this is doing with a switch, since the switch could return the same object this does.  So in that regard its probably just a preference. Though maybe this is marginally faster, since you can jump straight to the datasource you want rather than iterating a switch?

In either case, I would argue that the logic should still be in a util file, like this one.  
That way its re-usable and no code duplication is needed to call this elsewhere, and easily testable because there is no other code to have to mock out for these tests.   It also keeps the test code for the main calling classes smaller because they have one less thing to worry about.

NickPhura commented:
I pushed some changes to a second commit so its easier to review. Can squash them during the merge.

## Changes
- Removed default dateIssued from Orders and Inspections
- Removed an old enum file that was unused.
- Updated swagger to take an optional recordTypes array alongside the dataSourceType.
- Updated the epic-datasource.js to use the recordTypes array, if its not null.  Otherwise it defaults to updating all supported types.
- Updated the tests accordingly.

marklise commented on api/src/integrations/epic/epic-utils.js:
```diff
@@ -0,0 +1,44 @@
+
+  // Transform Epic author names to match values tracked by NRPTI
+  if (epicRecord.documentAuthor) {
+    switch (epicRecord.documentAuthor) {
```
Are these documentAuthor fields valid against the EPIC dataset?  I was under the impression that documentAuthor fields in EPIC are the actual names of people, or null, or empty string.  I can't validate this as I no longer have read access to EPIC DB.

NickPhura replied:
Ill be honest, this was a random update from Lena, that said to map epic author to a canned nrpti author value.  documentAuthor is the only value on the record, so I just assumed that was the one she meant.

Looking at the response though, it looks like documentAuthor is generally null, or an _id, so Ill have to ask her what she was thinking around this more in depth.

NickPhura commented on api/src/utils/task-audit-record.test.js:
```diff
@@ -4,7 +4,9 @@ describe('TaskAuditRecord', () => {
       // create mock create function
-      const mockCreateFunction = jest.fn(() => Promise.resolve({ _id: '123', created: true }));
+      const mockCreateFunction = jest.fn(() =>
+        Promise.resolve({ _id: '123', created: true, read: ['sysadmin'], write: ['sysadmin'] })
```
For the purpose of the test, it doesn't actually matter what we return, it could be a sentence even.  The real call to mongoose/mongo is being mocked out, so the ```_id: '123'``` is just there to provide evidence that the code called the mongoose function like we expected it to.  I only use _id because its recognizable and gives you some extra info about what kinds of things the test is doing.

NickPhura commented:
> I think it should be a set of parameters to pass in, or a predefined "ALL" param. I like the ability to stay finer grained than obscuring the details.

I'm not sure if this is what you meant, but this current version supports 2 ways to call the import:

1. ```startTask( { dataSourceType: 'epic' } )``` --> updates all record types for epic

2. ```startTask( { dataSourceType: 'epic', recordTypes: ['Order', 'Inspection'] } )``` --> updates only Order and Inspection types for epic.
